### ISSUE TITLE
Assign more UID range for regular users

### ISSUE THREAD

**Issue by wtmpx** [systemd/systemd#34617]
### Component

systemd

### Is your feature request related to a problem? Please describe

Improve: Nowadays companies often have more than 60,000 users/individuals. To enhance security, a common practice is to avoid reusing old user identification numbers (UIDs).

In current documentation/implementation we have below information (https://systemd.io/UIDS-GIDS/): 
```
1000…65533 and 65536…4294967294 → Everything else, i.e. regular (human) users.

Regular users: 1000–60000
Unused: 65536–524287
```

This create a conflict, because Unused doesn't mean Regular users. Unused means unused. This clarifies that the range 65534–524287 is indeed unused and not intended for regular users. Systemd might reserve this range for future use.

### Describe the solution you'd like

I would suggest increasing the range for regular users. Change: 
`Unused: 65536–524287 `
To be used all for Regular users or split it.

**Comment by poettering** [comment]
As discussed in that document, UIDs outside of the 16bit range are somewhat problematic, because they make it harder to use them inside of a container, under the assumption that each container gets 64K users assigned by default.

**Comment by poettering** [comment]
I think the best way out would be if the document would define a separate UID range for network/sssd/IPA style setups, to cover for these cases, and assign that to a range above 16bit. I proposed this before to the IPA people, but the categorically refused entertaining that idea, they insist that IPA should by default own the entire UID space, and not be confined to some range. I don't think a compromise is possible there. We just want clearly defined range assignments, they want to own everything, no solution in sight there. Sorry.

**Comment by wtmpx** [comment]
> they insist that IPA should by default own the entire UID space, and not be confined to some range. 
With current documentation/implementation, systemd have used some UIDs internally, e.g.: 

`61184…65519 | Dynamic service users | systemd | nss-systemd`

I am suggesting to document unused range. It would be nice to involve one of the IPA engineers in this issue.

**Comment by poettering** [comment]
So, I'd be happy if we say something like:

| Range      | Purpose
|------------|--------|
|65536…458751 | Networked/Organizatio-wide UID assignments for regular users (IPA, SSSD, LDAP, …) |
|458752…524287 | Unused |

that would leave almost 400K UIDs for organizations to assign. 

But the problem is that IPA folks don't like the idea, they want to basically own the whole UID range.

**Comment by poettering** [comment]
See https://github.com/systemd/systemd/pull/30846#issuecomment-1884506830 btw, for a whole thread with the IPA folks about this. It was a trainwreck. abbra is an IPA dude.

**Comment by wtmpx** [comment]
> Range	Purpose
> 65536…458751	Networked/Organizatio-wide UID assignments for regular users (IPA, SSSD, LDAP, …)

@poettering ,, I like this proposal. 
How can we involve Alexander Bokovoy in this request? would be very good to involve anyone from IPA folks in this request. 

useradd/usermod should comply with systemd UIDs and these tools should avoid creating human-users in reserved systemd UID rang. With current useradd,, below commands does create user with reserved systemd UID: 

# useradd --uid 60520 --gid 100 -m test8
# useradd --uid 61185 --gid 100 -m test9

>60514…60577	Host users mapped into containers
>61184…65519	Dynamic service users

**Comment by bb010g** [comment]
> See [#30846 (comment)](https://github.com/systemd/systemd/pull/30846#issuecomment-1884506830) btw, for a whole thread with the IPA folks about this. It was a trainwreck. abbra is an IPA dude.

The number that stuck out to me in that issue was jaltman's university that allocates individual UIDs to each of their 60\_000 (let's say 0xF000 = 61\_440) new students each year. He'd like for that to at least scale without recycling UIDs for 100 (let's say 0x80 = 128) years, which would mean 6 million (let's say 0x78\_0000 = 7\_864\_320) available regular UIDs minimum, with an expectation that containers work to some extent. I have no idea how many UIDs he expects to be _active_ from a machine's / workload's perspective at any given time, and thus mappable into a container. I'd expect exposing a network share mount that uses the whole regular UID range to a container wouldn't be feasible.

systemd currently expects mostly UIDs in \[0x0000, 0x1\_0000\) to be used. By doing so, `systemd-nspawn` can map blocks of 0x1\_0000 UIDs inside \[0x8\_0000, 0x7000\_0000\) for 0x6FF8 possible container blocks, leaving the 0x1000\_0000 UIDs (i.e. an extra 0x1000 blocks of 0x1\_0000 UIDs) in \[0x7000\_0000, 0x8000\_0000\) unused. If those unused UID blocks were to be used, the total number of container UID blocks would rise to 0x7FF8. The UID block size can't lower below 0x1\_0000 because of `nobody`. [IPA also assigns 0x1\_0000 subordinate UIDs to each user.](<https://freeipa.readthedocs.io/en/ipa-4-11/designs/subordinate-ids.html>).

I think it'd be good to establish a hard minimum for many containers systemd expects to have UID mapped on a system at once. (I don't expect the minimum to raise from 0x6FF8 due to the current space constraints.)

Looking at [FreeIPA's _Central management of subordinate user and group ids_ design documentation](<https://freeipa.readthedocs.io/en/ipa-4-11/designs/subordinate-ids.html>) again, FreeIPA limits regular UIDs to ≤0x8000_0000 and uses up to the whole range \[0x8000\_0000, 0xFFFF\_0008\) for subordinate UIDs. To then directly quote that document:

> A default subordinate id count of 65,536 and a total range of approx. 2.1 billion limits IPA to slightly more than 32,000 possible ranges. It may sound like a lot of users, but there are much bigger installations of IPA. For comparison Fedora Accounts has over 120,000 users stored in IPA.
>
> For that reason we treat subordinate id space as premium real estate and don’t auto-map or auto-assign subordinate ids by default. Instead we give the admin several options to assign them manually, semi-manual, or automatically.

And to quote systemd's documentation for that range:

> Note that the range 2147483648…4294967294 (i.e. 2^31…2^32-2) should be handled with care. Various programs (including kernel file systems — see `devpts` — or even kernel syscalls – see `setfsuid()`) have trouble with UIDs outside of the signed 32-bit range, i.e any UIDs equal to or above 2147483648. It is thus strongly recommended to stay away from this range in order to avoid complications. This range should be considered reserved for future, special purposes.

Subordinate UIDs might be an acceptable "special purpose" for that UID range. Whether systemd would want to move `systemd-nspawn`'s reservation from \[0x8\_0000, 0x7000\_0000\) (0x1000 blocks) to \[0x8000\_0000, 0xFFFF\_0008\) (0x7FFF blocks) is another matter. Moving `systemd-nspawn`'s full range would make \[0x1\_0000, 0x8000\_0000\) fully unused, which I'm sure the IPA folks would be happy about.

Given that IPA constrains itself to 0x7FFF total subordinate UID blocks, I think they'd at least have to be happy with systemd guaranteeing them 0x7FFF total regular UIDs. systemd currently guarantees them 0xE678 total regular UIDs from the range \[0x03E8, 0xEA60\], but [FreeIPA's defaults](<https://freeipa.readthedocs.io/en/ipa-4-11/designs/id-mapping.html>) leave UID range [0x0, 0x3\_0D40] to the local machine anyways, so these users take issue with the upper bound of the range [0x3\_0D41, 0x8\_0000\) (0x4\_F2BF total UIDs). If systemd guarantees a range of regular UIDs \[0x1\_0000, 0x7_0000\) (0x6\_0000 total UIDs) for networked users, then that's only a 21% increase (0x10D41 additional UIDs) over their effective current situation. That isn't much.

I really don't know what the IPA folks would be happy with besides leaving \[0x1\_0000, 0x8000\_0000\) unused. I think this is a symptom of a deeper problem, that numeric UIDs & GIDs are used far more in "portable" & archival data than they should be, so healthy recycling of numeric IDs can't occur. systemd-homed even tries to make UIDs stable, just at a _much_ smaller scale.

**Comment by wtmpx** [comment]
@bb010g thanks a lot for your comment, how can we drive this forward?

**Comment by sigprof** [comment]
> they insist that IPA should by default own the entire UID space, and not be confined to some range

If we were starting from scratch, we could define a range easily (e.g., 0x10000 ... 0x3FFFFFFF — IPA users, 0x40000000 ... 0x6FFFFFFF — containers, 0x70000000 ... 0x7FFFFFFF — reserved).  But there are already lots of IPA installs out there which may be using any IDs in the 0x30D40 (200000) ... 0x7738A13F (2000199999) range (and potentially even beyond that range if some custom install options were used, or a custom range was added after the initial install).  Do we want to declare about 40% of current IPA installs as permanently incompatible with systemd?  (Actually this may be more than 40%, because extra ID ranges are allocated when adding trusts, and those ranges also would have the same chance to be located above the 0x40000000 boundary, if we choose that value.)

(For the 0x40000000 ... 0x6FFFFFFF container ID allocation the IPA ID blocks 1 ... 5367 would be usable due to being entirely below that range, and blocks 9396 ... 10000 would be entirely above that range, so you would get 5367 + 605 = 5972 usable blocks, and 4028 unusable blocks.)

So it seems that any solution that places the container IDs in the 31-bit range and also does not break compatibility with existing IPA setups would require a way to configure the container ID range locally (using whatever way you like — some new parameters in `/etc/login.conf`, a new config file under `/etc/systemd`, …).

The alternative of using the high IDs (0x80000000 ... 0xFFFEFFFF — the 0xFFFF0000 ... 0xFFFFFFFF block cannot be used due to collision with `(uid_t)-1`) for containers is probably worth investigating.  However, this also should be coordinated with IPA, because they are also starting to use that region for subids allocation; maybe they could agree to change their defaults to use, e.g., the lower half of that region, and then systemd could use the upper half and start its allocations from the higher end of the region.

**Comment by wtmpx** [comment]
@keszybz how can we drive this request?

**Comment by bb010g** [comment]
> thanks a lot for your comment, how can we drive this forward?

Personally, I don't have long-term hope for any solution besides the Linux kernel introducing 128-bit UUIDs for UIDs/GIDs (which could probably be merged into one type at that point). I'd love to be proven wrong, but the more I read about this problem, the less optimistic I become.

If I could have one wish related to this problem granted, it would be for all relevant parties to come together and design a greenfield interface for identity in modern Linux.

**Comment by thkukuk** [comment]
The main problem is: there are very large networks out there with 150.000-200.000 users.
This networks are very big.
With old systemd versions (e.g. the one we shipped with SLES12), they could use the now reserved dynamic user range. And they do use this range, since the networks are older than systemd and the UID range was already used before.
Now they want to update to newer Linux distributions with a current systemd version and run into problems with user, who have an UID in that range.
It's impossible for them to move this users in another UID range, too many machines where files are stored with this UIDs.
But they e.g. don't care about containers, homed, nspawn or anything similar on this machines, at least not today.

Wouldn't it be possible to implement a config file, where the admin of such a network could specify the UID range for dynamic users?
For others nothing would change, but this users would get the change to update to a current systemd version.
Your opinion @poettering?

**Comment by wtmpx** [comment]
Till now we don't offer a clear guideline for end-user which ranges are allowed for normal users.

Looking again into https://systemd.io/UIDS-GIDS/, I see: 
> 1000…65533 and 65536…4294967294 → Everything else, i.e. regular (human) users.

however in above big range, systemd uses the following: 
> 524288…1879048191	Container UID ranges	systemd	nss-systemd 
> 2147483648…4294967294	HIC SVNT LEONES

Looking into unused ranges, we have these: 
> 60578…61183	Unused
> 65520…65533	Unused
> 65536…524287	Unused
> 1879048192…2147483647	Unused

And we have a statement about unused range: 
> Note that “Unused” in the table above doesn’t mean that these ranges are really unused. 
> It just means that these ranges have no well-established pre-defined purposes between Linux, 
> generic low-level distributions and systemd. There might very well be other packages that allocate from theseranges.

So, we cannot advice end-user to start using unused range for their normal users.
This range: 1000…60000	Regular users ,, is very small for big/old-established companies. 

@poettering can we enhance and rework the Ranges on above page.

**Comment by greskom** [comment]
Hello,

I do not fully understand this concept. We lived without it for decades. But if it is really needed, how about making the ranges configurable? For example I use kerberos/ldap setup for central user management and use UIDs starting from 1000000. Since the latest KDE Plasma updates the plasma-powerdevil service is not starting up, because all the centrally managed users are reported as container disposition. I do not know whether it would be better to add another disposition "central" or to add another range for regular users? I think this approach of configurable ranges could be usable not only for kerberos/ldap setups, but also for winbind or AD setups. Maybe this could be relevant also for IPA. I do not know whether the IPA administrator does have any control above UID assignment? If yes, it will not depend on IPA developers to accept some UID space partitioning.

Thanks

Marek

**Comment by wtmpx** [comment]
Can we reserve 10M from this 1800M range for normal users? 

**1879048192…2147352575 | 0x70000000…0x7FFDFFFF | 1879048192 | unused**

**Comment by gresko-ntt** [comment]
I still cannot understand why should not there exist a configuration file, where the administrator would be able to define his own UID plan. This would lead to all happy situation, because administrator usualy is aware of UID assignment in his environment.

**Comment by wtmpx** [comment]
Because above range still marked as unused. If admin start allocating users in above range and then this range was used by X technology in future and caused some conflict, admin need to migrate affected users to new UID. Current documentation says there is 60K normal user. This range should be larger.

**Comment by poettering** [comment]
> Hello,
> 
> I do not fully understand this concept. We lived without it for decades. But if it is really needed, how about making the ranges configurable? For example I use kerberos/ldap setup for central user management and use UIDs starting from 1000000. Since the latest KDE Plasma updates the plasma-powerdevil service is not starting up, because all the centrally managed users are reported as container disposition. 

This is your issue: your user management stack doesnt report a disposition hence as a fallback we derive it from the uid. If your user management stack would just report the right resolution then everything would work.

The disposition of the user is a property of the user record not of the system, hence a per-system config is simply the wrong place.

I suggested to the ipa/ssd folks a decade ago that they just provide complete records to us. The fact that they don't is just sad, but I think that's really on them.

**Comment by poettering** [comment]
> I still cannot understand why should not there exist a configuration file, where the administrator would be able to define his own UID plan. This would lead to all happy situation, because administrator usualy is aware of UID assignment in his environment.

If your user database stack of choice would just frickin report proper user records with dispositions it would be in control in saying what is what.

I am strongly against making the fallback disposition determination configurable, because it's a frickin' fallback. Just fix your fricking user databae stack to provide the right data and the issue goes away.

On some distros we actually do allow some minor configuration of certain uid boundaries, mostly to configure allocation ranges (i.e. where systemd allocates system users from in sysusers), not to change user disposition fallbacks so much.

**Comment by poettering** [comment]
> Can we reserve 10M from this 1800M range for normal users? 
> 
> **1879048192…2147352575 | 0x70000000…0x7FFDFFFF | 1879048192 | unused**
> 

Send a pr and we can discuss it. But it's not something that is an obvious choice because users allocated from that range cannot be used from inside of containers if 64k users are delegated to it (simply because those uids are above the 64k boundary). And in systemd at least we push container mgrs to delegate 64k. This must be discussed in the spec addition to some level.

Note that the correct fix of the fallback disposition classification of user records for this is that your user database stack starts reporting correct user records to systemd with a disposition field set. If they did that they could use pretty much any range they like and things work correctly.

The sad thing though is that even a decade or so after the infra for all this was added to sytemd ipa/ssd folks never showed too much interes in that.

**Comment by poettering** [comment]
And also of course the ipa folks never showed interest in allocating from such a uid subset by default anyway. So even if we define a second (high uid) range for regular users then it would remain a manual step to make ssd/ipa allocate from that

**Comment by greskom** [comment]
The IPA is not the only user management database which could be in use. there are many other solutions. For example winbind. Me personally I use LDAP with kerberos and I use UIDs starting from 1000000. I started using this solution prior to systemd born. I think systemd should take into account such solutions which are perfectly valid and widely used.

I do not fully understand you point. What do you mean by reporting correct user records to systemd? Where should I set disposition field? Should it be recorded as a LDAP attribute for the user? How would systemd read the record from LDAP? The configuration file seems the straightforward solution to me. There could be multiple ranges defined there. Why are you against? This would enable any administrator to adjust the range in systemd according to their deployed solution.

**Comment by bb010g** [comment]
> Note that the correct fix of the fallback disposition classification of user records for this is that your user database stack starts reporting correct user records to systemd with a disposition field set. If they did that they could use pretty much any range they like and things work correctly.

@poettering To clarify, the user database stack has to report individual, complete user records to systemd with the disposition field set? If a UID is not locally known, is known to be in the subordinate UID range, and cannot be fetched from the network currently, would that user database stack *not* be able to return a stub UID record that only indicates disposition and not whether that UID is currently allocated to a user (which cannot be known until a later fetch from the network)?

**Comment by poettering** [comment]
> I do not fully understand you point. What do you mean by reporting correct user records to systemd? Where should I set disposition field? Should it be recorded as a LDAP attribute for the user? How would systemd read the record from LDAP? The configuration file seems the straightforward solution to me. There could be multiple ranges defined there. Why are you against? This would enable any administrator to adjust the range in systemd according to their deployed solution.

Implement this: https://systemd.io/USER_RECORD/ and this: https://systemd.io/USER_GROUP_API/

It's simple, and you can report any disposition you want there.

**Comment by poettering** [comment]
> [@poettering](https://github.com/poettering) To clarify, the user database stack has to report individual, complete user records to systemd with the disposition field set? 

I am not sure what you mean by "complete"? systemd's user record logic knows a ton of fields. you don't have to fill them in if you don't want to. the most minimal record you have to provide is probably something like `{"userName":"lennart","uid":1000}`.

and if you want to report regular disposition it would be: 

`{"userName":"lennart","uid":1000,"disposition":"regular"}`

> If a UID is not locally known, is known to be in the subordinate UID range, and cannot be fetched from the network currently, would that user database stack _not_ be able to return a stub UID record that only indicates disposition and not whether that UID is currently allocated to a user (which cannot be known until a later fetch from the network)?

i don't grok this question, but the userdb api stuff is for querying/reporting existing users, it has no concept of uid allocation ranges if that's what you are asking.

**Comment by greskom** [comment]
> > I do not fully understand you point. What do you mean by reporting correct user records to systemd? Where should I set disposition field? Should it be recorded as a LDAP attribute for the user? How would systemd read the record from LDAP? The configuration file seems the straightforward solution to me. There could be multiple ranges defined there. Why are you against? This would enable any administrator to adjust the range in systemd according to their deployed solution.
> 
> Implement this: https://systemd.io/USER_RECORD/ and this: https://systemd.io/USER_GROUP_API/
> 
> It's simple, and you can report any disposition you want there.

Are you kidding me? I have central user management in ldap. Should I create another json file and distribute it on all the machines around? If there is implementation of things like this it should be able to integrate it to ldap. I still see the configurable user uid ranges as a fast viable solution. Why are you so strongly against?

**Comment by poettering** [comment]
> Are you kidding me? I have central user management in ldap. Should I create another json file and distribute it on all the machines around?

No, certainly not. It's an api your local ldap client is supposed to implement that's all, it just needs to generate these short json objects from its ldap data source to tell systemd about the metadata of its user records. All that should happen on-the-fly whenever a request comes in from systemd. These are not supposed to ever be stored on disk as-is if they come from some other data source.

This has been that way for a bunch of years btw, it's not that hard.

**Comment by greskom** [comment]
Even better, so the administrator should develop software to provide some api for the system to work correctly?

**Comment by sigprof** [comment]
> > If a UID is not locally known, is known to be in the subordinate UID range, and cannot be fetched from the network currently, would that user database stack _not_ be able to return a stub UID record that only indicates disposition and not whether that UID is currently allocated to a user (which cannot be known until a later fetch from the network)?
> 
> i don't grok this question, but the userdb api stuff is for querying/reporting existing users, it has no concept of uid allocation ranges if that's what you are asking.

Looks like this implies that systemd should get some new configuration files or APIs for UID allocation ranges.

Currently IPA can potentially allocate UIDs anywhere in the 200000…2000199999 (0x00030D40…0x7738A13F) range, which completely covers the systemd container UID range (0x00080000…0x6FFFFFFF) and extends beyond it into the currently unused 0x00010000…0x0007FFFF and 0x70000000…0x7FFDFFFF ranges.  However, in practice IPA would be using some number of contiguous blocks of 200000 UIDs in that range for its allocations, and the IPA client should be able to get the list of those blocks (and cache that information so that it would be available even when the IPA server is unreachable).  The missing part is the way for the IPA client to report those allocations to systemd, so that the systemd code which allocates UID blocks for containers could avoid overlaps with those blocks.

There could be more complications; e.g., if IPA has an AD trust with `ipa-ad-trust-posix`, UID/GID values for AD users would come directly from the `uidnumber`/`gidnumber` AD attributes without any translation, and in that case the used ranges won't be declared anywhere.  So there still could be a need for local configuration of excluded ID ranges (or not really local, if those files would be managed by some kind of central management software).

If those new configuration files would also provide a way to specify the default user disposition for UIDs within the described blocks, this might actually be enough to make things work with existing IPA clients without rewriting the client code to provide full JSON user records.

**Comment by poettering** [comment]
> Even better, so the administrator should develop software to provide some api for the system to work correctly?

you developed your ldap client yourself? if so, then yes, implement the varlink api. if you didn't and somebody else, then ask them to just implement the interfaces. it's been a decade. if they don't, find another project instead? but maybe look for one that friendly towards modern linux, i.e. systemd, and just implements the trivial apis necessary.

**Comment by greskom** [comment]
Hello,

I use standard openldap setup on Fedora with sssd. Is there something included in the Fedora what can be used as an API? If not how exactly should the API work? I understand what should be the returned json, but I have no clue what should be the API request and where API lies? How would systemd contact the API? Should it be some socket in /run/systemd/userdb directory? Should it be somehow registered?

Thanks

Marek

**Comment by greskom** [comment]
To be more specific, the LDAP interconnection is not necessary. It would suffice for me for example to do simple nss request. If user x belongs to group users, it is regular user. Is this easily achievable?

**Comment by poettering** [comment]
> If those new configuration files would also provide a way to specify the default user disposition for UIDs within the described blocks, this might actually be enough to make things work with existing IPA clients without rewriting the client code to provide full JSON user records.

sorry, but i am not convinced that a configuration file is the way to go to fix such random UID/GID assignments. User records should be self descriptive, and the designator-by-uid-range thing is just a fallback for cases where a correct designator is not available. Just provide proper (i.e. userdb json) records for existing records, with the right disposition set. And in future make sure to allocate from one of the already spec'ed UID ranges, and things will be better. (And if the spec'ed UID range is too small for some setups, as mentioned, we can add another range, somebody would need to prep a patch for that).

Providing proper user records is *easy*, it's just a matter of implementing a small IPC iface that has been part of most Linux platforms for many years.

**Comment by poettering** [comment]
> I use standard openldap setup on Fedora with sssd. Is there something included in the Fedora what can be used as an API? If not how exactly should the API work? I understand what should be the returned json, but I have no clue what should be the API request and where API lies? How would systemd contact the API? Should it be some socket in /run/systemd/userdb directory? Should it be somehow registered?

The docs for this have been posted already: https://systemd.io/USER_GROUP_API/

**Comment by greskom** [comment]
> The docs for this have been posted already: https://systemd.io/USER_GROUP_API/

The docs is completely unreadable for me.

Problem 1: If the sockets are examined in parallel, what if io.systemd.NameServiceSwitch returns the result sooner as my socket? I will get false record.

Problem 2: There is only some python like description of some methods. I would not use python, since I am not familiar with. If I create C program to serve the socket, how should the protocol look like?

**Comment by sigprof** [comment]
> sorry, but i am not convinced that a configuration file is the way to go to fix such random UID/GID assignments.

Some kind of range API is still needed to avoid collisions when allocating ID ranges for containers.  E.g., suppose that systemd decided to allocate 0x1e0000…0x1effff (1966080…2031615) for some container.  However, IPA may be allocating IDs from a range like 1800000…1999999, or maybe 2000000…2199999, so the 1966080…2031615 range is not actually available.  Just probing the first and last IDs in the range won't help, because those particular IDs might not be allocated by IPA at the probe time; even probing all 65536 IDs (which would probably be prohibitively expensive) won't guarantee that IPA won't allocate a new ID in that range just after the probe.  Fixing this issue properly would require some API to report ID ranges which are not available for container ID allocations (or any other allocations, if systemd gets more features which want a large number of UIDs/GIDs).  And the IPA client would also need to cache those ranges, so that the case when a container is started while the system is offline is handled as well as it could be.  The mentioned configuration file could be considered a simple implementation of such API and local cache.

(There still would be a chance of collision when IPA admins add a new ID range; but this could be mitigated by preparing the new ID range before the old one is depleted completely, so that any containers started before the new ID range was added would die off before the range is actually used.)

**Comment by greskom** [comment]
> The docs for this have been posted already: https://systemd.io/USER_GROUP_API/

I tried to use chatgpt to produce some example. It was of some use to start to understand the docs. But I found one huge problem. I need only to override the disposition parameter. But it seems it is needed to return complete record for the user I want to change the disposition for. Is it somehow possible only to override? It is not feasible to monitor whole the structure that should be needed to define for the user. Moreover when it could change over time.

**Comment by poettering** [comment]
> > The docs for this have been posted already: https://systemd.io/USER_GROUP_API/
> 
> I tried to use chatgpt to produce some example. It was of some use to start to understand the docs. But I found one huge problem. I need only to override the disposition parameter. But it seems it is needed to return complete record for the user I want to change the disposition for. Is it somehow possible only to override? It is not feasible to monitor whole the structure that should be needed to define for the user. Moreover when it could change over time.

hmm, just return the record data NSS returns too, but with that one additional field filled in?

**Comment by greskom** [comment]
> hmm, just return the record data NSS returns too, but with that one additional field filled in?

Hmm, I tried to return:


    sd_json_build(&record, SD_JSON_BUILD_OBJECT(
        SD_JSON_BUILD_PAIR("userName", SD_JSON_BUILD_STRING(pw->pw_name)),
        SD_JSON_BUILD_PAIR("uid", SD_JSON_BUILD_UNSIGNED(pw->pw_uid)),
        SD_JSON_BUILD_PAIR("gid", SD_JSON_BUILD_UNSIGNED(pw->pw_gid)),
        SD_JSON_BUILD_PAIR("realName", SD_JSON_BUILD_STRING(pw->pw_gecos)),
        SD_JSON_BUILD_PAIR("homeDirectory", SD_JSON_BUILD_STRING(pw->pw_dir)),
        SD_JSON_BUILD_PAIR("shell", SD_JSON_BUILD_STRING(pw->pw_shell)),
        SD_JSON_BUILD_PAIR("disposition", SD_JSON_BUILD_STRING(disposition))
    ));


but this was apparently not enough.

Would it be possible to override only?
